### PR TITLE
compute: add timely_step_duration_seconds histogram

### PR DESCRIPTION
**Pull request by danhhz**
In incident 70, we found that timely-driven Futures were not promptly being polled after being woken. This was presumably because some operator was not regularly yielding. Not yielding also causes issues in query interactivity (i.e. because peeks are included in this), so it's also otherwise undesirable. The introspection sources have full-fidelity data around operator run times, but we duplicate a proxy for them into prometheus so we can alert on it and easily scan it in investigations.

Concretely, a new histogram is added for step_or_park durations in compute timely clusters. This doesn't tell us anything about individual operators (and including the park time is undesirable) but it has the benefit of being very easy yet still useful by virtue of being an upper bound. In particular, anything over a second or two might be a hint to investigate.

Touches https://github.com/MaterializeInc/cloud/issues/7602

### Motivation

  * This PR adds a feature that has not yet been specified.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/compute/src/metrics.rs [hunk 1]
```diff
@@ -69,6 +82,12 @@ impl ComputeMetrics {
                 help: "Whether maintenance is currently occuring.",
                 var_labels: ["worker_id"],
             )),
+            timely_step_duration_seconds: registry.register(metric!(
+                name: "mz_timely_step_duration_seconds",
+                help: "The time spent in each compute step_or_park call",
+                const_labels: {"cluster" => "compute"},
+                buckets: mz_ore::stats::histogram_seconds_buckets(0.000_128, 32.0),
```
Comment (teskje) [inline src/compute/src/metrics.rs:89]: I was about to suggest having this metric also for the storage cluster, but based on the name and the label you already thought of that. Did you plan to make that a separate PR?

**Review by teskje — APPROVED** [review]
Thanks! This looks straightforward to me.

Looking at the implementation of `step_or_park` there indeed doesn't seem to be a way to differentiate between a step and a park. I thought we could call `worker.activations.borrow().empty_for()` ourselves before the call to see if the worker would park, but `step_or_park` also mutates `activations` before making that check, so that would probably not work.

File: src/compute/src/metrics.rs (hunk 1 above)
Comment (danhhz) [reply src/compute/src/metrics.rs:89]: I hadn't looked at it yet, tbh. Wanted to make sure this one made the cutoff for next week's release candidate, so quickly typed it up and that's how far I've gotten. Follow up PR makes sense to me

**Review by danhhz — COMMENTED** [review]
TFTR!

> Looking at the implementation of step_or_park there indeed doesn't seem to be a way to differentiate between a step and a park. I thought we could call worker.activations.borrow().empty_for() ourselves before the call to see if the worker would park, but step_or_park also mutates activations before making that check, so that would probably not work.

I also thought about calling `step_or_park` first with a zero duration (which doesn't park) and measuring that, and then calling `step_or_park` with the `None`, but 1) it's racy with persist_source activations and 2) it's a big enough change to how things work that the risk didn't feel worth it
